Result/Result.swift

```diff
@@ -179,6 +179,15 @@ public func ?? <T, Error> (left: Result<T, Error>, @autoclosure right: () -> Res
 	return left.recoverWith(right())
 }
 
+// MARK: - Derive result from failable closure
+
+public func materialize<T>(f: () throws -> T) -> Result<T, ErrorType> {
```

Comment:
I don't know if this would work, but maybe we can provide another function for when the caller knows the type of error being thrown? Something like:

```swift
public func materialize<T, E: ErrorType>(f: () throws -> T) -> Result<T, E> {
	do {
		return .success(try f())
	} catch {
		return .failure(error as E)
	}
}
```